Index: trunk/src/plugins/splash/details/plugin.c
===================================================================
--- trunk.orig/src/plugins/splash/details/plugin.c
+++ trunk/src/plugins/splash/details/plugin.c
@@ -344,10 +344,8 @@
   }
   else
     write_on_views (plugin,
-                    "Password",
-                    strlen ("Password"));
-
-  write_on_views (plugin, ":", strlen (":"));
+                    "Password:",
+                    strlen ("Password:"));
 
   for (i = 0; i < bullets; i++)
     write_on_views (plugin, "*", strlen ("*"));
